libbeat/processors/actions/alterFieldProcessor.go

```diff
@@ -0,0 +1,145 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package actions
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/processors"
+	conf "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+type alterFieldProcessor struct {
+	Fields         []string
+	IgnoreMissing  bool
+	FailOnError    bool
+	AlterFullField bool
+
+	processorName string
+	alterFunc     mapstr.AlterFunc
+}
+
+// NewAlterFieldProcessor is an umbrella method for processing events based on provided fields. Such as converting event keys to uppercase/lowercase
+func NewAlterFieldProcessor(c *conf.C, processorName string, alterFunc mapstr.AlterFunc) (beat.Processor, error) {
+	config := struct {
+		Fields         []string `config:"fields"`
+		IgnoreMissing  bool     `config:"ignore_missing"`
+		FailOnError    bool     `config:"fail_on_error"`
+		AlterFullField bool     `config:"alter_full_field"`
+	}{
+		IgnoreMissing:  false,
+		FailOnError:    true,
+		AlterFullField: true,
+	}
+
+	if err := c.Unpack(&config); err != nil {
+		return nil, fmt.Errorf("failed to unpack the %s fields configuration: %w", processorName, err)
+	}
+
+	// Skip mandatory fields
+	for _, readOnly := range processors.MandatoryExportedFields {
+		for i, field := range config.Fields {
+			if field == readOnly {
```

Comment:
I suspect this condition needs to be inverted. If the field on the configured list matches the fields on the mandatory list, we should skip it, right?

Also, we cannot simply check for matching here, we should check if the configured field is not a prefix to the mandatory field when we compare in a case-insensitive manner.

We need tests for this code.

```suggestion
			if !strings.HasPrefix(strings.ToLower(readOnly), strings.ToLower(field)) {
```

`ToLower` calls can be done beforehand, so we don't `ToLower` the same `readOnly` string on each inner iteration.